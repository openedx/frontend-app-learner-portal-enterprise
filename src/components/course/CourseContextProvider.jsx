import { createContext } from 'react';

export const CourseContext = createContext();
