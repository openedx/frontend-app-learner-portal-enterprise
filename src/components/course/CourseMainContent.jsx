import React, { useContext } from 'react';
import { breakpoints, Hyperlink, MediaQuery } from '@edx/paragon';
import { AppContext } from '@edx/frontend-platform/react';

import { PreviewExpand } from '../preview-expand';
import { CourseContext } from './CourseContextProvider';
import CourseSidebar from './CourseSidebar';
import CreatedBy from './CreatedBy';
import VerifiedCertPitch from './VerifiedCertPitch';

function formatSponsorTextList(sponsors) {
  const names = sponsors.map(sponsor => sponsor.name);
  let sponsorTextList;

  if (names.length === 1) {
    [sponsorTextList] = names;
  }

  if (names.length === 2) {
    sponsorTextList = names.join(' and ');
  }

  if (names.length > 2) {
    const lastName = names.pop();
    sponsorTextList = `${names.join(', ')}, and ${lastName}`;
  }

  return sponsorTextList;
}

const CourseMainContent = () => {
  const { config } = useContext(AppContext);
  const { state } = useContext(CourseContext);
  const { course, activeCourseRun } = state;

  return (
    <>
      <MediaQuery minWidth={breakpoints.large.minWidth}>
        {matches => !matches && (
          <div className="mb-5">
            <CourseSidebar />
          </div>
        )}
      </MediaQuery>
      {course.fullDescription && (
        <PreviewExpand
          className="mb-5"
          cta={{
            labelToExpand: 'More about this course',
            labelToMinimize: 'Collapse about this course',
            id: 'about-this-course',
          }}
          heading={<h3>About this course</h3>}
        >
          {/* eslint-disable-next-line react/no-danger */}
          <div dangerouslySetInnerHTML={{ __html: course.fullDescription }} />
        </PreviewExpand>
      )}
      {course.sponsors && course.sponsors.length > 0 && (
        <div className="mb-5">
          <h3>Sponsored by</h3>
          <div className="row no-gutters mt-3">
            {course.sponsors.map((sponsor) => (
              <div className="col-lg-6 mb-3" key={sponsor.name}>
                <div className="mb-2">
                  <a
                    href={`${config.MARKETING_SITE_BASE_URL}/${sponsor.marketingUrl}`}
                    aria-hidden="true"
                    tabIndex="-1"
                    target="_blank"
                    rel="noopener noreferrer"
                  >
                    <img src={sponsor.logoImageUrl} alt={`${sponsor.name} logo`} />
                  </a>
                </div>
                <Hyperlink
                  destination={`${config.MARKETING_SITE_BASE_URL}/${sponsor.marketingUrl}`}
                  target="_blank"
                >
                  {sponsor.name}
                </Hyperlink>
              </div>
            ))}
          </div>
          <p>
            The production of this course would not have been possible without the
            generous contributions of {formatSponsorTextList(course.sponsors)}.
          </p>
        </div>
      )}
      {course.outcome && (
        <PreviewExpand
          className="mb-5"
          cta={{
            labelToExpand: 'Expand what you\'ll learn',
            labelToMinimize: 'Collapse what you\'ll learn',
            id: 'what-youll-learn',
          }}
          heading={<h3>What you&apos;ll learn</h3>}
        >
          {/* eslint-disable-next-line react/no-danger */}
          <div dangerouslySetInnerHTML={{ __html: course.outcome }} />
        </PreviewExpand>
      )}
      {course.syllabusRaw && (
        <PreviewExpand
          className="mb-5"
          cta={{
            labelToExpand: 'Expand syllabus',
            labelToMinimize: 'Collapse syllabus',
            id: 'course-syllabus',
          }}
          heading={<h3>Syllabus</h3>}
        >
          {/* eslint-disable-next-line react/no-danger */}
          <div dangerouslySetInnerHTML={{ __html: course.syllabusRaw }} />
        </PreviewExpand>
      )}
      <CreatedBy />
      {activeCourseRun.type === 'verified' && (
        <VerifiedCertPitch />
      )}
      {course.learnerTestimonials && (
        <div className="mb-5">
          <h3>Learner testimonials</h3>
          {/* eslint-disable-next-line react/no-danger */}
          <div dangerouslySetInnerHTML={{ __html: course.learnerTestimonials }} />
        </div>
      )}
      {course.faq && (
        <div className="mb-5">
          <h3>Frequently asked questions</h3>
          {/* eslint-disable-next-line react/no-danger */}
          <div dangerouslySetInnerHTML={{ __html: course.faq }} />
        </div>
      )}
      {course.additionalInformation && (
        <div className="mb-5">
          {/* eslint-disable-next-line react/no-danger */}
          <div dangerouslySetInnerHTML={{ __html: course.additionalInformation }} />
        </div>
      )}
      {activeCourseRun.hasOfacRestrictions && (
        <div className="mb-5">
          <h3>Who can take this course?</h3>
          <p>
            Unfortunately, learners from one or more of the following countries or regions will not
            be able to register for this course: Iran, Cuba and the Crimea region of Ukraine.
            While edX has sought licenses from the U.S. Office of Foreign Assets Control (OFAC) to
            offer our courses to learners in these countries and regions, the licenses we have
            received are not broad enough to allow us to offer this course in all locations. EdX
            truly regrets that U.S. sanctions prevent us from offering all of our courses to
            everyone, no matter where they live.
          </p>
        </div>
      )}
    </>
  );
};

export default CourseMainContent;
