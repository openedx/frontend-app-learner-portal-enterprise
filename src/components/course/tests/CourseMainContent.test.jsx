import renderer from 'react-test-renderer';
import { AppContext } from '@edx/frontend-platform/react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import CourseMainContent from '../CourseMainContent';
import { CourseContext } from '../CourseContextProvider';

jest.mock('../CourseSidebar', () => jest.fn(() => <div data-testid="course-sidebar" />));
jest.mock('../../preview-expand', () => ({
  PreviewExpand: jest.fn(({ children }) => (
    <div data-testid="preview-expand">
      {children}
    </div>
  )),
}));

jest.mock('../CreatedBy', () => jest.fn(() => <div data-testid="created-by" />));

jest.mock('../VerifiedCertPitch', () => jest.fn(() => <div data-testid="verified-cert-pitch" />));

const baseCourseContextValue = {
  state: {
    activeCourseRun: {
      type: 'verified',
      hasOfacRestrictions: true,
    },
    course: {
      fullDescription: '<p>Test Description</p>',
      sponsors: [{
        name: 'Test Sponsor',
        marketingUrl: 'https://test.org',
        logoImageUrl: 'https://test.org/logo.png',
      }],
      outcome: '<p>Test Outcome</p>',
      syllabusRaw: '<p>Test Syllabus</p>',
      learnerTestimonials: '<p>Test Testimonials</p>',
      faq: '<p>Test FAQ</p>',
      additionalInformation: '<p>Test Additional Information</p>',
    },
  },
};

const baseAppContextValue = {
  config: {
    MARKETING_SITE_BASE_URL: 'https://test.org',
  },
};

const CourseMainContentWrapper = ({
  appContextValue = baseAppContextValue,
  courseContextValue = baseCourseContextValue,
}) => (
  <AppContext.Provider value={appContextValue}>
    <CourseContext.Provider value={courseContextValue}>
      <CourseMainContent />
    </CourseContext.Provider>
  </AppContext.Provider>
);

describe('ExternalCourseEnrollment', () => {
  it('renders and matches snapshot', () => {
    const tree = renderer.create(<CourseMainContentWrapper />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handles 2 sponsors', () => {
    const courseContextValue = {
      ...baseCourseContextValue,
      state: {
        ...baseCourseContextValue.state,
        course: {
          ...baseCourseContextValue.state.course,
          sponsors: [{
            name: 'Test Sponsor',
            marketingUrl: 'https://test.org',
            logoImageUrl: 'https://test.org/logo.png',
          }, {
            name: 'Another Sponsor',
            marketingUrl: 'https://test.org',
            logoImageUrl: 'https://test.org/logo.png',
          }],
        },
      },
    };
    render(<CourseMainContentWrapper courseContextValue={courseContextValue} />);
    expect(screen.getByText('Test Sponsor and Another Sponsor', { exact: false })).toBeInTheDocument();
  });

  it('handles 3+ sponsors', () => {
    const courseContextValue = {
      ...baseCourseContextValue,
      state: {
        ...baseCourseContextValue.state,
        course: {
          ...baseCourseContextValue.state.course,
          sponsors: [{
            name: 'Test Sponsor',
            marketingUrl: 'https://test.org',
            logoImageUrl: 'https://test.org/logo.png',
          }, {
            name: 'Another Sponsor',
            marketingUrl: 'https://test.org',
            logoImageUrl: 'https://test.org/logo.png',
          }, {
            name: 'Plus One Sponsor',
            marketingUrl: 'https://test.org',
            logoImageUrl: 'https://test.org/logo.png',
          }],
        },
      },
    };
    render(<CourseMainContentWrapper courseContextValue={courseContextValue} />);
    expect(screen.getByText('Test Sponsor, Another Sponsor, and Plus One Sponsor', { exact: false })).toBeInTheDocument();
  });
});
