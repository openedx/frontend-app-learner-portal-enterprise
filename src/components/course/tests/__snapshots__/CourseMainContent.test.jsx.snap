// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExternalCourseEnrollment renders and matches snapshot 1`] = `
Array [
  <div
    className="mb-5"
  >
    <div
      data-testid="course-sidebar"
    />
  </div>,
  <div
    data-testid="preview-expand"
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Test Description</p>",
        }
      }
    />
  </div>,
  <div
    className="mb-5"
  >
    <h3>
      Sponsored by
    </h3>
    <div
      className="row no-gutters mt-3"
    >
      <div
        className="col-lg-6 mb-3"
      >
        <div
          className="mb-2"
        >
          <a
            aria-hidden="true"
            href="https://test.org/https://test.org"
            rel="noopener noreferrer"
            tabIndex="-1"
            target="_blank"
          >
            <img
              alt="Test Sponsor logo"
              src="https://test.org/logo.png"
            />
          </a>
        </div>
        <a
          className="pgn__hyperlink default-link standalone-link"
          href="https://test.org/https://test.org"
          onClick={[Function]}
          rel="noopener noreferrer"
          target="_blank"
        >
          Test Sponsor
          <span
            className="pgn__hyperlink__external-icon"
            title="Opens in a new tab"
          >
            <span
              className="pgn__icon"
              style={
                Object {
                  "height": "1em",
                  "width": "1em",
                }
              }
            >
              <svg
                aria-hidden={true}
                fill="none"
                focusable={false}
                height={24}
                role="img"
                viewBox="0 0 24 24"
                width={24}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 19H5V5h7V3H3v18h18v-9h-2v7zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3h-7z"
                  fill="currentColor"
                />
              </svg>
              <span
                className="sr-only"
              >
                in a new tab
              </span>
            </span>
          </span>
        </a>
      </div>
    </div>
    <p>
      The production of this course would not have been possible without the generous contributions of 
      Test Sponsor
      .
    </p>
  </div>,
  <div
    data-testid="preview-expand"
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Test Outcome</p>",
        }
      }
    />
  </div>,
  <div
    data-testid="preview-expand"
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Test Syllabus</p>",
        }
      }
    />
  </div>,
  <div
    data-testid="created-by"
  />,
  <div
    data-testid="verified-cert-pitch"
  />,
  <div
    className="mb-5"
  >
    <h3>
      Learner testimonials
    </h3>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Test Testimonials</p>",
        }
      }
    />
  </div>,
  <div
    className="mb-5"
  >
    <h3>
      Frequently asked questions
    </h3>
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Test FAQ</p>",
        }
      }
    />
  </div>,
  <div
    className="mb-5"
  >
    <div
      dangerouslySetInnerHTML={
        Object {
          "__html": "<p>Test Additional Information</p>",
        }
      }
    />
  </div>,
  <div
    className="mb-5"
  >
    <h3>
      Who can take this course?
    </h3>
    <p>
      Unfortunately, learners from one or more of the following countries or regions will not be able to register for this course: Iran, Cuba and the Crimea region of Ukraine. While edX has sought licenses from the U.S. Office of Foreign Assets Control (OFAC) to offer our courses to learners in these countries and regions, the licenses we have truly regrets that U.S. sanctions prevent us from offering all of our courses to received are not broad enough to allow us to offer this course in all locations. EdX everyone, no matter where they live.
    </p>
  </div>,
]
`;
